### PR TITLE
Added `PartitionedWriter` for disk partitioning.

### PR DESCRIPTION
**Pull request by illumination-k**
related #2257

It seems difficult to customize IpcWriter to archive disk partitioning because disk partitioning needs to create directories and write IPC to files multiple times.
Therefore, I added a new struct for this purpose.

Usage is like the following.

```rust
use polars_core::prelude::*;
use polars_core::df;
use polars_io::ipc::PartitionIpcWriter;
let df = df!("a" => [1, 1, 2, 3], "b" => [2, 2, 3, 4], "c" => [2, 3, 4, 5]).unwrap();
let by = ["a", "b"];
PartitionIpcWriter::new("./root_dir", by)
    .finish(&df)
```

and result like below

```
root_dir/
  a=1/
    b=2/
      part-cf737804-90ea-4c37-94f8-9aa016f6953a.ipc
  a=2/
    b=3/
      part-cf737804-90ea-4c37-94f8-9aa016f6953a.ipc
  a=3/
    b=4/
      part-cf737804-90ea-4c37-94f8-9aa016f6953a.ipc
```

I'm not sure which style (e.g., hive) is best for disk partitioning, so any opinions would be appreciated.

### CODE REVIEW

File: polars/polars-io/src/ipc.rs [hunk 1]
```diff
@@ -274,6 +273,115 @@ where
...
+
+    pub fn finish(self, df: &DataFrame) -> Result<()> {
+        use polars_core::POOL;
+        use rayon::iter::IntoParallelIterator;
+        use rayon::iter::ParallelIterator;
+        use uuid::Uuid;
+
+        let rootdir: PathBuf = self.rootdir.into();
+        let by: Vec<String> = self.by.into_iter().map(|x| x.as_ref().to_owned()).collect();
+        let group = df.groupby(&by)?;
```
Comment (ritchie46) [inline polars/polars-io/src/ipc.rs:338]: We can use existing `partition_by` operation here.

```rust
        let partitioned = df.partition_by(by)?;
        partitioned.iter().for_each(|df| {
            path.push(format!("part-{}.ipc", Uuid::new_v4()));

            let mut file = std::fs::File::create(path)?;
            let writer = BufWriter::new(&mut file);

            IpcWriter::new(writer)
                .with_compression(self.compression)
                .finish(&mut partition_df)
        });

```

File: polars/polars-io/src/ipc.rs [hunk 2]
```diff
@@ -274,6 +273,115 @@ where
...
+                    .get_groups()
+                    .idx_ref()
+                    .into_par_iter()
+                    .map(|t| {
+                        let mut partition_df =
+                            unsafe { df.take_iter_unchecked(t.1.iter().map(|i| *i as usize)) };
+                        let mut path = resolve_partition_dir(&rootdir, &by, &partition_df);
+                        std::fs::create_dir_all(&path)?;
+
+                        path.push(format!("part-{}.ipc", Uuid::new_v4()));
```
Comment (ritchie46) [inline polars/polars-io/src/ipc.rs:352]: Why would we want a UUID here instead of a count?

**Review by ritchie46 — COMMENTED** [review]
The overall idea looks good. 

I do think we must generalize this in a `PartitonedWriter<F>` which then accepts on of our existing writers. Maybe we must add a trait and implement it for our writers. In that case we can reuse the partitioned logic for all writers.

**Comment by illumination-k** [comment]
Agree.

I'm currently considering the following design

```rust
trait WriterFactory {
  fn create_writer(&self, writer: Box<dyn Writer>) -> Box<dyn SerWriter>;
}

struct IpcWriterOptions {
  compression: Option<compression>
}

impl WriterFactory for IpcWriterOptions {
  fn create_writer(&self, writer: Box<dyn Writer>) -> Box<dyn SerWriter> {
     Box::new(IpcWriter::new(writer).with_compression(self.compression))
  }
}

struct ParitionedWriter<F> {
  factory: F
}

impl<F> PartitionedWriter<F: WriterFactory> {
  fn finish(self, df: &mut DataFrame) -> Result<()> { 
    ...
    let mut file = std::fs::File::create(path)?;
    let writer = BufWriter::new(&mut file);
    self.factory.create_writer(Box::new(writer)).finish(&mut partitioned_df)?;
    ...
    Ok(())
  }
}
```
This design allows applying partitioned logic to our writers by implementing each writer option.

for this design, we need to consider #3333.

What do you think?

File: polars/polars-io/src/ipc.rs (hunk 1 above)
Comment (illumination-k) [reply polars/polars-io/src/ipc.rs:338]: I know `partitioned_by` function, but this function collects all of the DataFrames. I think this logic does not collect all, so reducing memory allocation.

Comment (ritchie46) [reply polars/polars-io/src/ipc.rs:338]: Ok, I think we should then adapt `partition_by` a little bit by letting that return `Result<impl Iterator<Item=DataFrame>>`. That way we can reduce the amount of code and keep it more DRY.

Comment (illumination-k) [reply polars/polars-io/src/ipc.rs:338]: Cool. It seems the best way.

File: polars/polars-io/src/ipc.rs (hunk 2 above)
Comment (illumination-k) [reply polars/polars-io/src/ipc.rs]: There is no need to use UUID, but some other libraries use it. Using a count makes sense.

**Comment by illumination-k** [comment]
I refactored following the above design.

closes #3333

File: polars/polars-io/src/ipc.rs (hunk 1 above)
Comment (illumination-k) [reply polars/polars-io/src/ipc.rs]: I used `partition_by` this time

**Comment by universalmind303** [comment]
I'm wondering, how will this affect the read operations? Will we need to create a partition aware reader to fully benefit from disk partitions.

If we write to a partitioned directory, then do a scan operation with a filter, we should be able to pushdown that query to the disk read operation & skip files entirely based on certain predicates. 

for example. 

```py
df = pl.DataFrame({'col_1': ["A","A", "B", "B"], "col_2": [1,2,3,4]})
df.partitioned_write("/path/to/ipc/dir", by=["col_1"])

# ideally, we would know that the files are partitioned by 'col_1'
#  so we can skip reading `/path/to/ipc/dir/col_1=B*` entirely. 
df = pl.scan_ipc("/path/to/ipc/dir/*").filter(pl.col("col_1") == pl.lit("A"))..collect()

**Comment by ritchie46** [comment]
Thanks @illumination-k looks good!

**Pull request merged**